pkg/controller/apiextensions/composite/composed/api.go

```diff
@@ -51,17 +51,23 @@ type DefaultConfigurator struct{}
 
 // Configure applies the raw template and sets name and generateName.
 func (*DefaultConfigurator) Configure(cp resource.Composite, cd resource.Composed, t v1alpha1.ComposedTemplate) error {
-	// Any existing name will be overwritten when we unmarshal the template. We
-	// store it here so that we can reset it after unmarshalling.
+	// Any existing name/namespace will be overwritten when we unmarshal the
+	// template. We store it here so that we can reset it after unmarshalling.
 	name := cd.GetName()
+	namespace := cd.GetNamespace()
 	if err := json.Unmarshal(t.Base.Raw, cd); err != nil {
 		return errors.Wrap(err, errUnmarshal)
 	}
-	// Unmarshalling the template will overwrite any existing fields, so we must
-	// restore the existing name, if any. We also set generate name in case we
-	// haven't yet named this composed resource.
+	// If composite resource was created via a requirement, it will already have
+	// a random string appended. Appending yet another random string increases
+	// the odds that the composed managed resource won't be provisioned because
+	// of long name that the provider wouldn't accept.
 	cd.SetGenerateName(cp.GetName() + "-")
+	if cp.GetRequirementReference() != nil && cp.GetRequirementReference().Name != "" {
+		cd.SetGenerateName(cp.GetRequirementReference().Name + "-")
+	}
```

Comment:
Agree that this feels a little weird to sometimes name it one way and sometimes name it another.